### PR TITLE
[AIR-894] Making network plugins CR singleton

### PR DESCRIPTION
**Pull request by joey00072**
Jira: [AIR-894](https://platform9.atlassian.net/browse/AIR-894)

  - networkplugins CR cluster scoped
  - single instance of network plugin is allowed on cluster

Dependency:
=> https://github.com/platform9/luigi/pull/90  assume this is merged
[cert-manager](https://cert-manager.io/docs/installation/#default-static-install) need to be installed on cluster

### Testing

Screenshot 2023-07-05 at 8 09 40 PM
Screenshot 2023-07-05 at 7 40 36 PM

[AIR-894]: https://platform9.atlassian.net/browse/AIR-894?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: controllers/networkplugins_webhook.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	if multusExist && (networkPluginsReq.Spec.Plugins == nil || networkPluginsReq.Spec.Plugins.Multus == nil) {
+		if resp, err := a.multusUninstallCheck(); err != nil {
+			log.Error(err, "error while doing checks for multus uninstall")
+			return resp
+		}
+
+	}
+
+	if err := a.isNetworkPluginsValid(networkPluginsReq, networkPluginsList); err != nil {
+		log.Error(err, "Networking Plugins already exist, New can not be installed before removing old ones")
```
Comment (mithilarun) [inline controllers/networkplugins_webhook.go:80]: NIT
```suggestion
		log.Error(err, "NetworkPlugins already exists")
```

File: controllers/networkplugins_webhook.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+		return admission.Denied(fmt.Sprintf("Networking Plugins already exists: %v", err.Error()))
```
Comment (mithilarun) [inline controllers/networkplugins_webhook.go:81]:
```suggestion
		return admission.Denied(fmt.Sprintf("NetworkPlugins already exists: %v", err.Error()))
```

**Review by mithilarun — COMMENTED** [review]
Minor NITS inline. Also, I see a lot of overlapping code with https://github.com/platform9/luigi/pull/90, please merge that first and rebase.

How do we plan on installing cert-manager? Is that going to be bundled with Luigi?

File: controllers/networkplugins_webhook.go (hunk 2 above)
Comment (joey00072) [reply controllers/networkplugins_webhook.go]: Suggestions Commited

File: README.md [hunk 3]
```diff
@@ -101,3 +101,8 @@ spec:
 ```
 
 That is it! Now that you have the secondary CNIs and other related plugins deployed, you may need to prep the nodes before you can actually create Multus Networks and assign them to Pods. In order to do so, use Luigi's own HostPlumber plugin: https://github.com/platform9/luigi/blob/master/hostplumber/README.md
+
+
+##### Dev note
+This project need to migrate to Kubebuilder/v4.
+weekhooks where added manually `make generate && make manifestes` will not add required feild for webhook in crds and luigi deployment. refer `samples/luigi-plugins-operator-v2.yaml`
```
Comment (poojaghumre) [inline README.md:108]: nit: typos

File: README.md [hunk 4]
```diff
@@ -101,3 +101,8 @@ spec:
 ```
 
 That is it! Now that you have the secondary CNIs and other related plugins deployed, you may need to prep the nodes before you can actually create Multus Networks and assign them to Pods. In order to do so, use Luigi's own HostPlumber plugin: https://github.com/platform9/luigi/blob/master/hostplumber/README.md
+
+
+##### Dev note
+This project need to migrate to Kubebuilder/v4.
```
Comment (poojaghumre) [inline README.md:107]: Is moving to v4 still required?

File: samples/luigi-plugins-operator-v2.yaml
Comment (poojaghumre) [inline samples/luigi-plugins-operator-v2.yaml:1]: Is this file just a backup of the older kubebuilder-v2 yaml and for kubebuilder-v3, its samples/luigi-plugins-operator.yaml?

If yes, just updating the 2nd file should suffice.

File: README.md (hunk 4 above)
Comment (joey00072) [reply README.md:107]: v3 is deprecated 2 years ago, weebhook that we added do not scaffold automatically.
its in patched working state now. 
ideal to migrate
https://kubebuilder.io/migration/v3vsv4.html

File: samples/luigi-plugins-operator-v2.yaml
Comment (joey00072) [reply samples/luigi-plugins-operator-v2.yaml:1]: samples/luigi-plugins-operator-v2.yaml This is new yaml with webhook

**Review by poojaghumre — APPROVED** [review]

File: samples/luigi-plugins-operator-v2.yaml [hunk 5]
```diff
@@ -388,7 +399,7 @@ spec:
         - --leader-elect
         command:
         - /manager
-        image: docker.io/platform9/luigi-plugins:private-master-shubhamc-weebhook-multus-pmk-2738082
+        image: docker.io/platform9/luigi-plugins:private-master-shubhamc-AIR-895-singleton-pmk-2739563
```
Comment (poojaghumre) [inline samples/luigi-plugins-operator-v2.yaml:402]: Just noticed this.. you will need to revert this private image

**Pull request merged**